File: material/algorithm/sort/TopoSort.js
/**
 * 拓扑排序-Kahn排序
 * 先将依赖转化为有向图
 * 通过有向图，能获取出来图的每一个点所指向的下一个点。
 * 创建一个map，记录每一个点被指向这个点的数量。
 * 创建一个队列qunue，将前面map中被指向这个点数量为0的点放入这个队列。
 * 将这个队列的点进行出队
 * 每个点出队的时候，这个点的指向的所有的点的被指向这个点的数量减1。就是map中记录的点（出队点的指向点）数量减一，
 * 如果数量减为0了（这个点的前置点都出队列了），将这个点也加入出队队列。
 */
let dependce = [
  'f->g',
  'c->g',
  'd->e',
  'b->d',
  'c->e',
  'b->c',
  'a->f',
  'f->c'
]
class Graph {
  constructor () {
    this.linkedList = {}
    this.size = 0
  }
  add (s, t) {
    if (!this.linkedList[s]) {
      this.linkedList[s] = []
      this.size++
    }
    this.linkedList[s].push(t)
  }
}

let g = new Graph()
for (let i = 0; i < dependce.length; i++) {
  let d = dependce[i].split('->')
  g.add(d[0], d[1])
}

function TopoSortByKahn (graph) {
  let inDegree = {}
  for (let g in graph.linkedList) {
    if (inDegree[g] === undefined) {
      inDegree[g] = 0
    }
    let arr = graph.linkedList[g]
    for (let i = 0; i < arr.length; i++) {
      let g1 = arr[i]
      if (inDegree[g1] === undefined) {
        inDegree[g1] = 0
      }
      inDegree[g1]++
    }
  }
  let queue = []
  for (let key in inDegree) {
    if (inDegree[key] === 0) {
      queue.push(key)
      delete inDegree[key]
    }
  }
  let array = []
  while(queue.length > 0) {
    let i = queue.shift()
    array.push(i)
    let g = graph.linkedList[i]
    for (let j = 0; g && j < g.length; j++) {
      let g1 = graph.linkedList[i][j]
      inDegree[g1]--
      if (inDegree[g1] === 0) {
        queue.push(g1)
        delete inDegree[g1]
      }
    }
  }
  console.log(array)
}

TopoSortByKahn(g)

File: material/algorithm/sort/TopoSort.java
package material.algorithm.sort;

import java.util.LinkedList;

public class TopoSort {
  public static void main(String[] args) {
    TopoSort ts = new TopoSort();
    ts.graph(7);
    // ts.addEdge(4, 5);
    ts.addEdge(5, 6);
    ts.addEdge(2, 6);
    ts.addEdge(3, 4);
    ts.addEdge(1, 3);
    ts.addEdge(2, 4);
    ts.addEdge(1, 2);
    ts.addEdge(0, 5);
    ts.addEdge(5, 2);
    ts.topoSortByKahn();
  }
  private int v;
  private LinkedList<Integer> adj[];
  public void graph (int v) {
    this.v = v;
    this.adj = new LinkedList[v];
    for (int i = 0; i < v; ++i) {
      adj[i] = new LinkedList<>();
    }
  }
  public void addEdge(int s, int t) {
    adj[s].add(t);
  }
  
  public void topoSortByKahn() {
    int[] inDegree = new int[v]; // 统计每个顶点的入度
    for (int i = 0; i < v; ++i) {
      for (int j = 0; j < adj[i].size(); ++j) {
        int w = adj[i].get(j); // i->w
        // System.out.print("-->" + w);
        inDegree[w]++;
      }
    }
    LinkedList<Integer> queue = new LinkedList<>();
    for (int i = 0; i < v; ++i) {
      if (inDegree[i] == 0)
        queue.add(i);
    }
    while (!queue.isEmpty()) {
      int i = queue.remove();
      System.out.print("->" + i);
      for (int j = 0; j < adj[i].size(); ++j) {
        int k = adj[i].get(j);
        inDegree[k]--;
        if (inDegree[k] == 0)
          queue.add(k);
      }
    }
  }
}
